core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp

```diff
@@ -130,63 +126,78 @@ inline void check_shmem_request(CudaInternal const* cuda_instance, int shmem) {
   }
 }
 
-// This function needs to be template on DriverType and LaunchBounds
+// These functions needs to be template on DriverType and LaunchBounds
 // so that the static bool is unique for each type combo
 // KernelFuncPtr does not necessarily contain that type information.
+
+template <class DriverType, class LaunchBounds, class KernelFuncPtr>
+const cudaFuncAttributes& get_cuda_kernel_func_attributes(
+    const KernelFuncPtr& func) {
+  // Race condition inside of cudaFuncGetAttributes if the same address is
+  // given requires using a local variable as input instead of a static Rely
+  // on static variable initialization to make sure only one thread executes
+  // the code and the result is visible.
+  auto wrap_get_attributes = [func]() -> cudaFuncAttributes {
+    cudaFuncAttributes attr_tmp;
+    KOKKOS_IMPL_CUDA_SAFE_CALL(cudaFuncGetAttributes(&attr_tmp, func));
+    return attr_tmp;
+  };
+  static cudaFuncAttributes func_attr = wrap_get_attributes();
+  return func_attr;
+}
+
 template <class DriverType, class LaunchBounds, class KernelFuncPtr>
-inline void configure_shmem_preference(KernelFuncPtr const& func,
-                                       bool prefer_shmem) {
+inline void configure_shmem_preference(const KernelFuncPtr& func,
+                                       const cudaDeviceProp& device_props,
+                                       const size_t block_size,
+                                       const size_t shmem,
+                                       const size_t occupancy) {
 #ifndef KOKKOS_ARCH_KEPLER
+  const auto& func_attr =
+      get_cuda_kernel_func_attributes<DriverType, LaunchBounds>(func);
+  const size_t maxshmem_per_sm   = device_props.sharedMemPerMultiprocessor;
+  const size_t shmem_per_block   = shmem + func_attr.sharedSizeBytes;
+  const size_t maxthreads_per_sm = device_props.maxThreadsPerMultiProcessor;
+  // Limits due do registers/SM
+  const size_t regs_per_sm     = device_props.regsPerMultiprocessor;
+  const size_t regs_per_thread = func_attr.numRegs;
+  // The granularity of register allocation is chunks of 256 registers per warp
+  // -> 8 registers per thread
+  const size_t allocated_regs_per_thread = 8 * ((regs_per_thread + 8 - 1) / 8);
+  const size_t max_blocks_regs =
+      regs_per_sm / (allocated_regs_per_thread * block_size);
+  const size_t numthreads_desired =
+      ((maxthreads_per_sm * (100 / occupancy) + 31) / 32) * 32;
```

Comment:
Why is it 100/occupancy and not occupancy/100?

Reply:
because I was stupid